### ISSUE TITLE
Internal error with `OSSL_DECODER_fetch`

### ISSUE THREAD

**Issue by manison** [openssl/openssl#17456]
I'm trying to get familiar with the new concept around `OSSL_DECODER` introduced in OpenSSL 3 and maybe I'm doing something wrong but I have suspicion of a bug in `OSSL_DECODER_fetch` API.

Following is the minimal repro:

```c
#include <openssl/provider.h>
#include <openssl/decoder.h>

void main(void)
{
	OSSL_LIB_CTX *libctx = NULL;
	OSSL_PROVIDER *defprov = OSSL_PROVIDER_load(libctx, "default");
	OSSL_DECODER *decoder = OSSL_DECODER_fetch(libctx, "der", "provider=default,input=pem");
}
```

The call `OSSL_DECODER_fetch` fails with debug assertion

> crypto\property\property.c:637: OpenSSL internal error: Assertion failed: prov != NULL

and program terminates (in debug build).

This is because `inner_ossl_decoder_fetch` calls `ossl_method_store_cache` [(here)](https://github.com/openssl/openssl/blob/openssl-3.0.1/crypto/encode_decode/decoder_meth.c#L395-L396) with the second parameter (`OSSL_PROVIDER *prov`) set to `NULL`. However the function `ossl_method_store_cache` does not expect `NULL` here and throws an assertion [(here)](https://github.com/openssl/openssl/blob/openssl-3.0.1/crypto/property/property.c#L637).

I'm using custom built openssl-3.0.1.

**Comment by mattcaswell** [comment]
This does appear to be a problem and is a regression since 3.0.0 which does not have this behaviour.

The problem was introduced in PR #16725. I'm unsure why this doesn't result in test failures in the decoder tests. From code inspection it looks like this same problem will also impact encoders and store loaders.

@levitte

**Comment by mattcaswell** [comment]
> I'm unsure why this doesn't result in test failures in the decoder tests

The answer to this is that it is actually fairly unusual to fetch an individual decoder. The normal way of using the decoder APIs is to set up an `OSSL_DECODER_CTX` and then call `OSSL_DECODER_from_bio` or similar function to iterate through the available decoders to find the right one for the data that we are trying to decode.

That said there *is* still a test which is explicitly there to test that fetching an individual decoder works, and its here:

https://github.com/openssl/openssl/blob/328bf5adf9e23da523d4195db309083aa02403c4/test/provfetchtest.c#L222-L283

Given the reproducer above we might have expected this test to fail as well. Stepping through it in a debugger reveals that the key difference is the presence of a query string. This test doesn't use one but the reproducer above does. That turns out to be significant since the failing codepath is only hit if a query string is present.

**Comment by mattcaswell** [comment]
Fix in #17459
